java/java.mx.project/src/org/netbeans/modules/java/mx/project/SuiteSources.java

```diff
@@ -0,0 +1,1195 @@
+    private final List<Library> libraries;
+    private final List<Dist> distributions;
+    private final FileObject dir;
+    /**
+     * non-null if the dependencies haven't yet been properly initialized
+     */
+    private Map<String, Dep> transitiveDeps;
+    /**
+     * avoid GC of imported projects
+     */
+    private final SuiteProject prj;
+    private final Map<String, SuiteSources> imported;
+
+    SuiteSources(SuiteProject owner, FileObject dir, MxSuite suite) {
+        final Map<String, Dep> fillDeps = new HashMap<>();
+        this.prj = owner;
+        this.dir = dir;
+        this.groups = findGroups(fillDeps, suite, dir);
+        this.libraries = findLibraries(fillDeps, suite);
+        this.imported = findImportedSuites(dir, suite, fillDeps);
+        this.distributions = findDistributions(suite, this.libraries, this.groups, fillDeps);
+        this.suite = suite;
+        this.transitiveDeps = fillDeps;
+    }
+
+    @Override
+    public String toString() {
+        return "MxSources[" + (dir == null ? "mx" : dir.toURI()) + "]";
+    }
+
+    private List<Group> findGroups(Map<String, Dep> fillDeps, MxSuite s, FileObject dir) {
+        List<Group> arr = new ArrayList<>();
+        for (Map.Entry<String, MxProject> entry : s.projects().entrySet()) {
+            String name = entry.getKey();
+            MxProject mxPrj = entry.getValue();
+            FileObject prjDir = findPrjDir(dir, name, mxPrj);
+            if (prjDir == null) {
+                fillDeps.put(name, new Group(name, mxPrj, null, null, null, name, name));
+                continue;
+            }
+            String prevName = null;
+            Group firstGroup = null;
+            String binPrefix;
+            if (mxPrj.subDir() == null) {
+                binPrefix = "mxbuild/";
+            } else {
+                binPrefix = "mxbuild/" + mxPrj.subDir() + "/";
+            }
+            for (String rel : mxPrj.sourceDirs()) {
+                FileObject srcDir = prjDir.getFileObject(rel);
+                FileObject binDir = getSubDir(dir, binPrefix + name + "/bin");
+                FileObject srcGenDir = getSubDir(dir, binPrefix + name + "/src_gen");
+                if (srcDir != null && binDir != null) {
+                    String prgName = name + "-" + rel;
+                    String displayName;
+                    if (prevName == null) {
+                        displayName = name;
+                    } else {
+                        displayName = name + "[" + rel + "]";
+                    }
+                    Group g = new Group(name, mxPrj, srcDir, srcGenDir, binDir, prgName, displayName);
+                    arr.add(g);
+                    if (firstGroup == null) {
+                        firstGroup = g;
+                    }
+                    prevName = displayName;
+                }
+            }
+            if (firstGroup != null) {
+                fillDeps.put(name, firstGroup);
```

Comment:
Silly question, but why just the 1st group ?

Reply:
Running the tests I haven't seen a case where there was more groups than `firstGroup`. But I don't want to replace it with `assert` now.